File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-0.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {

        ListNode dummy(0);
        ListNode *tail = &dummy;

        ListNode* l1 = list1;
        ListNode* l2 = list2;

        while (l1 != nullptr && l2 != nullptr) {
            if(l1->val < l2->val){
                tail->next = l1;
                l1 = l1->next;
            }else{
                tail->next = l2;
                l2 = l2->next;
            }

            tail = tail->next;

        }

        if(l1 != nullptr) tail->next = l1;
        else tail->next = l2;


        return dummy.next;

        
    }
};

File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:    
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        import heapq
        heap = []
        dummy = ListNode(0)
        curr = dummy
        
        for i , node in enumerate(lists):
            if node:
                heapq.heappush(heap , (node.val , i , node))
            
        
        while(heap):
            val , i , node = heapq.heappop(heap)
            curr.next = node
            if node.next:
                heapq.heappush(heap , (node.next.val , i ,node.next))
            curr = curr.next


        return dummy.next


        
        